Filter documentation (#116)
AY2122S1-CS2103T-T11-2/tp: docs/DeveloperGuide.md

### Common classes

Classes used by multiple components are in the `lingogo.commons` package.

--------------------------------------------------------------------------------------------------------------------

## **Implementation**

This section describes some noteworthy details on how certain features are implemented.

### Filter feature

#### Implementation

The filter feature is facilitated by `ModelManager`. It extends `Model` and implements `updateFilteredFlashcardList` 
which returns an unmodifiable view of filtered flashcards in the GUI.

The filter feature also relies on a nested `FilterBuilder` class within `FilterCommand`. Multiple filters can be 
given by the user in one command, however only one predicate (filter) can be accepted by 
`Model::updateFilteredFlashcList` to produce the filtered flashcards. `FilterBuilder` helps by generating multiple 
predicates from the user input, and then combining them into a single predicate. 

`FilterBuilder` is also a mutable class which allows processed user inputs to be directly set as variables within a 

> `FilterBuilder` instance. A mutable design is acceptable for `FilterBuilder` since it acts as a throwaway variable, 
> with only a one-time usage within `FilterCommand`. Furthermore, setVariable methods in `FilterBuilder` reduces the 

Reviewer: ```suggestion
`FilterBuilder` instance. A mutable design is acceptable for `FilterBuilder` since it only has a one-time usage within `FilterCommand`. Furthermore, setVariable methods in `FilterBuilder` reduces the 
```
